dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java

```diff
@@ -1364,9 +1372,86 @@ private void addRawData( DataQueryParams params, Grid grid )
+
+        for ( List<Object> row : grid.getRows() )
+        {
+            if ( isInBasePeriod( basePeriods, (String) row.get( 1 ) ) )
+            {
+                // Key is composed of [uid-period]
+                final String key = StringUtils.join(
+                    ArrayUtils.remove( row.toArray( new Object[0] ), grid.getWidth() - 1 ),
+                    DimensionalObject.DIMENSION_SEP );
+
+                final DimensionalItemObject dimensionalItemObject = findDimensionalItem( (String) row.get( 0 ), items ).get( 0 );
+                DimensionalItemObject clone = dimensionalItemObject;
+                if ( dimensionalItemObject.getPeriodOffset() != 0 )
+                {
+                    List<Object> periodOffsetRow = getPeriodOffsetRow( grid, dimensionalItemObject, (String)row.get(1), dimensionalItemObject.getPeriodOffset());
+                    if ( periodOffsetRow != null )
+                    {
+                        result.put( key,
+                            new DimensionItemWithValue( dimensionalItemObject,
+                                (Double) periodOffsetRow.get( grid.getWidth() - 1 ) ) );
+
+                    } // TODO throw exception?
+                    clone = SerializationUtils.clone( dimensionalItemObject );
+                    clone.resetPeriodOffset();
+                }
+
+                result.put( key,
+                    new DimensionItemWithValue( clone, (Double) row.get( grid.getWidth() - 1 ) ) );
+            }
+        }
+        
+        return result;
+    }
+    
+    private List<Object> getPeriodOffsetRow( Grid grid, DimensionalItemObject dimItem, String isoPeriod, int offset )
+    {
+        Period shifted = shiftPeriod( PeriodType.getPeriodFromIsoString( isoPeriod ), offset );
+        for ( List<Object> row : grid.getRows() )
+        {
+            final String rowUid = (String) row.get(0);
+            final String rowPeriod = (String) row.get( 1 );
+            if ( rowUid.equals(dimItem.getUid()) && rowPeriod.equals( shifted.getIsoDate() ) )
+            {
+                return row;
+            }
+        }
+        return null;
+
+    }
+    
+    private boolean isInBasePeriod( List<DimensionalItemObject> basePeriods, String period )
+    {
+        for ( DimensionalItemObject dimensionalItemObject : basePeriods )
+        {
+            Period p = (Period) dimensionalItemObject;
+            if ( p.getIsoDate().equals( period ) )
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Filters by uid and returns one ore more {@see DimensionalItemObject} from a
+     * List
+     * 
+     * @param uid a uid to filter {@see DimensionalItemObject} on
+     * @param items the filtered List
+     * @return a List only containing the  {@see DimensionalItemObject} matching the uid
+     */
+    private List<DimensionalItemObject> findDimensionalItem( String uid, List<DimensionalItemObject> items )
```

Comment:
Could we move this method to a utility class, add Javadoc and add unit test? E.g. `DimensionalObjectUtils`.

Reply:
Fix + test